Support video in mtmd (base API)

MrAMS opened the pull request:
**Purpose: Add foundational video support to `mtmd`**

This PR introduces the basic infrastructure required to support video processing in `mtmd`.

The primary motivation is to enable new models like `minicpm-v 4.5`, which require a processing technique that handles multiple video frames at once. In these models, the frames are no longer independent during the "eval chunk" step.

**Implementation:**
- **FFmpeg Integration:** Leverages FFmpeg as a new optional dependency to decode video files and extract individual frames.
- **`nz` Dimension:** Adds a flexible third dimension (`nz`) to the mtmd `bitmap` structure to support this new multi-frame data with minimal changes.
-  **Explicit Video Type:** Introduces explicit video-related flags to allow the API to distinguish a "sequence of video frames" from a "batch of unrelated images."
-  **Streaming Foundation:** Adds a `DecodedFramesQueue` to manage frames, laying the groundwork for future streaming scenarios (e.g., a producer/consumer model).

This PR serves as the necessary foundation for our follow-up work, which will implement the full `minicpm-v 4.5` multi-frame processing technique.

tc-mb committed:
add video

Signed-off-by: tc-mb <caitianchi@modelbest.cn>

MrAMS committed:
feat: support minicpm-v video normal speed

MrAMS committed:
fix: video normal speed uniform sample & check videoInfo valid

MrAMS committed:
refactor: support video in media

MrAMS committed:
refactor: add bitmap_type, instead of is_audio, is_video...

MrAMS committed:
feat: support loading video from buffer

MrAMS committed:
feat&fix: fix segmentation fault bug, add llama-server video_url api

MrAMS committed:
fix: max video frames

MrAMS committed:
clean up

MrAMS committed:
refactor: get_video_info and decode_video use the same AVFormatContext

MrAMS committed:
Merge upstream/master

ngxson commented on tools/mtmd/mtmd-helper.cpp:
```diff
@@ -436,25 +446,97 @@ mtmd_bitmap * mtmd_helper_bitmap_init_from_buf(mtmd_context * ctx, const unsigne
+           lower.rfind(".png")  != std::string::npos ||
+           lower.rfind(".bmp")  != std::string::npos ||
+           lower.rfind(".gif")  != std::string::npos ||
+           lower.rfind(".webp") != std::string::npos;
```
I'm not convinced by the fact that we should check for file extension. This is quite prone to error.

We can't we check by magic bytes instead?

MrAMS replied:
Thank you for the review! You are right. I will push the requested changes soon.

MrAMS replied:
This issue has already been fixed in [commit c5e9151](https://github.com/tc-mb/llama.cpp/commit/c5e9151e660aef48f8565739367864a1cca9312d).

ngxson commented on tools/mtmd/clip.cpp:
```diff
@@ -3639,16 +3640,67 @@ struct llava_uhd {
-            res.overview_size = clip_image_size{slice_size, slice_size};
-            res.refined_size  = clip_image_size{0, 0};
-            res.grid_size     = clip_image_size{0, 0};
+        if (clip_is_minicpmv(ctx)) {
```
What's the change here compared to the initial version?

IMO I think this is more like a duplicated logic, see `clip_image_preprocess` on how this is used by minicpm-v

ngxson commented on tools/mtmd/mtmd-video.cpp:
If the sole purpose of this file is to convert video into images, it should be outside of the core mtmd library.

The core library only need to care about what is an image. It doesn't care what is a video. If we get a video, we can use `mtmd-helper` (which is outside of the core) to convert video into images

ngxson reviewed (changes requested):
I expect the implementation to be much more simple: the mtmd doesn't have to know anything about video, but instead each frame is supplied via a separated bitmap. There will be no changes to mtmd API at all. You can add a helper to convert video to bitmap, it should be inside `mtmd-helper` to stay outside of the core library.

As a reminder, `mtmd-helper` is an optional unit, outside of the core `libmtmd`. It can only interface with `libmtmd` via the public API. For example, for audio processing, loading and resampling mp3/wav/flac files is done at `mtmd-helper` level, and the final PCM is fed into `libmtmd`. There is no resampling or decoding logic inside `libmtmd`.

Also, unless you do temporal pooling (like what Qwen does), there is no need to modify `clip.cpp`

So overall, I unfortunately can't accept the current approach. However, I'm available to help if we want to simplify this.

MrAMS commented on tools/mtmd/mtmd.h:
```diff
@@ -120,13 +124,19 @@ MTMD_API int mtmd_get_audio_bitrate(mtmd_context * ctx);
 MTMD_API uint32_t              mtmd_bitmap_get_nx     (const mtmd_bitmap * bitmap);
 MTMD_API uint32_t              mtmd_bitmap_get_ny     (const mtmd_bitmap * bitmap);
 MTMD_API const unsigned char * mtmd_bitmap_get_data   (const mtmd_bitmap * bitmap);
+MTMD_API unsigned char * mtmd_bitmap_get_data_mutable (mtmd_bitmap * bitmap);
```
I implemented it this way to avoid the significant overhead from copying large video data.

My approach, as seen in `convert_frames_to_bitmap`, is to first allocate the uninitialized bitmap by passing nullptr to `mtmd_bitmap_init_from_video`. Then, I get a mutable pointer using `mtmd_bitmap_get_data_mutable` and populate the data in place.

I'm definitely open to feedback, so if you have a suggestion for a better or more idiomatic approach, I would greatly appreciate the guidance!

MrAMS replied:
Thank you for sharing your perspective. I understand your concern that adding the `mtmd_bitmap_get_data_mutable` interface might introduce design overhead or deviate from the established coding conventions.

My personal view was that the performance gain from avoiding the extra memory copy justified this trade-off.

However, if you feel that the cost of the memory copy is more acceptable than the cost of compromising the design principle, I am happy to modify it as you requested.

ngxson replied:
Hmm, are you sure that this function will eliminate copy?

This is the place where you use it:

```cpp
    for (size_t i = 0; i < files.size(); i++) {
        if (i % stride != 0) continue;
        const std::string & f = files[i];
        if (!mtmd_helper::has_image_ext(f)) continue;
        mtmd::bitmap bmp(mtmd_helper_bitmap_init_from_file(ctx, f.c_str()));
        if (!bmp.ptr) continue;
        if(loaded==0){
            w = bmp.nx();
            h = bmp.ny();
            out_frames = mtmd_bitmap_init_from_video(w, h, nframes, nullptr);
            dest = mtmd_bitmap_get_data_mutable(out_frames);
        }
        GGML_ASSERT(bmp.nx() == w && bmp.ny() == h); // all frames must have the same size
        std::memcpy(dest,
                    bmp.data(),
                    bmp.n_bytes());
        dest += bmp.n_bytes();
        loaded++;
        if (loaded >= nframes) break;
    }
```

Notice that `std::memcpy`, you're copying the memory.

What I suggest is to instead of copying the memory, allocate multiple `mtmd_bitmap_init_from_video` like this:

```cpp
unsigned char * frame_offset = nx*ny*i; // i is the index of current frame in the video
mtmd_bitmap_init_from_video(..., frame_data + frame_offset);
```

ngxson replied:
I'll think more about how the API should be extended to support video input, will let you know when I'm ready.

> This technique requires handling multiple video frames at once

Currently for minicpm-v 4.5, how many frames are being processed together?

MrAMS replied:
> Hmm, are you sure that this function will eliminate copy?

You can check the code in `convert_frames_to_bitmap`; the copy overhead is avoided there.

The overhead is not avoided here because it reuses the `mtmd_helper_bitmap_init_from_file function`. And the overhead here is only a single frame's copy overhead, which is small compared to the video copy overhead within `convert_frames_to_bitmap`.

MrAMS replied:
> The overhead is not avoided here because it reuses the `mtmd_helper_bitmap_init_from_file function`. And the overhead here is only a single frame's copy overhead, which is small compared to the video copy overhead within `convert_frames_to_bitmap`.

Maybe I can change `AV_PIX_FMT_RGBA` to `AV_PIX_FMT_RGB24` in `decode_video_ffmpeg_to_rgba_from_format_ctx` to aviod RGBA->RGB :thinking:

ngxson commented:
> * You are right to point out the new video dependency. This was introduced because minicpm-v 4.5 requires a new processing technique to speed up video understanding. It needs to handle multiple video frames at once, as the frames are no longer independent during the "eval chunk" step, which requires holistic information from the set of frames.

In this case, I think it's better to firstly figure out how to re-enable real batching for certain models (probably starts with minicpm-v). I removed batching support because it was tricky to maintain, but we will eventually need to bring it back. Batching is needed if we want to do pooling multiple frames at once.

I think pooling multiple frame at once will be pretty much the norm for video processing for future models, so it's worth to support it. The video can be broken down into groups of frames, for example if the video contains 10 frames and the model support temporal pooling of 2, we will end up having 10/2=5 chunks.

Btw, could you point me to the python reference code of minicpm-v 4.5, where multiple frames are merged together?

MrAMS commented:
> In this case, I think it's better to firstly figure out how to re-enable real batching for certain models (probably starts with minicpm-v). I removed batching support because it was tricky to maintain, but we will eventually need to bring it back. Batching is needed if we want to do pooling multiple frames at once.

I agree that supporting multi-frame processing is the correct goal. Regarding the "batching" approach, I was slightly concerned that a generic batching mechanism might make the API's intent ambiguous—specifically, how to distinguish between a "batch of unrelated images" and a "sequence of video frames," which require different handling.

To keep the code clear and simple for this specific video use-case, I added the flexible third dimension `nz` and `video` type to the bitmap. My thinking was that this is a very minimal-cost modification that directly supports the video requirement.

Moreover, the meaning of `nz` itself can be further extended, which lays a foundation for future extensibility. This approach solves our immediate need for video frame processing clearly, without complicating the separate, larger discussion of re-enabling general-purpose batching.

ehoogeveen-medweb commented:
Possibly relevant, are the batches treated separately or does the model use something closer to a sliding window (potentially requiring images from 2 consecutive batches)?

ngxson commented on tools/mtmd/mtmd.h:
```diff
@@ -112,6 +112,10 @@ MTMD_API bool mtmd_support_audio(mtmd_context * ctx);
 
+// set MiniCPM-V UHD slicing upper bound (used when preprocessing images for MiniCPM-V)
+// values < 1 will be clamped to 1
+MTMD_API void mtmd_set_minicpmv_max_slice_nums(mtmd_context * ctx, int n);
```
Just want to note that we never accept creating model-specific public API like this one.

Also, number of slices (as well as the slicing process) is completely managed by `mtmd_tokenize`, the should never leak this info to the public API.

MrAMS committed:
undo the changes and submit a separate pull request to the llama-server later

MrAMS committed:
refactor: modifications made based on reviews

MrAMS committed:
Merge remote-tracking branch 'upstream/master' into Support-video-in-mtmd

MrAMS committed:
fix: add queue header file

ngxson commented:
> @tc-mb and I have re-confirmed, and the code in `tools/mtmd/clip.cpp` (specifically `minicpmv_max_slice_nums` and the `if (clip_is_minicpmv(ctx))` check) is necessary for our use case. The reason is that the minicpmv model requires `minicpmv_max_slice_nums` as a global, model-wide parameter. This parameter is _also_ required by the `get_slice_instructions` function to correctly process its image slicing logic.

I don't mean `clip_is_minicpmv` is not required, but I think you should put it outside of `get_slice_instructions`. You can see the example in `clip_image_preprocess` where we supply parameters to `get_slice_instructions`. The main goal is to make `get_slice_instructions` a generic, model-independent function.

---

Since I still need the second PR to fully understand how the API and the code base should be structured, I think it's safer if you can just **push everything to this PR**. There is no need to create a 2nd PR.

**Please note that I won't merge your PR**, but in parallel I will make a more cleaner version (like https://github.com/ngxson/llama.cpp/pull/32) which will eventually be merged to the main project. This way, you can freely implement what you want. I can make the code looks good, and we can also save our time from having to go back and forth between us.

I hope you understand this, but as the core maintainer of the library, I have the responsibility to maintain a code base that will work with all kinds of models: Qwen, MiniCPM, Llama, Gemma, etc. Therefore, I cannot risk adding too many model-specific functionalities - they should be generalize to work cross multiple models. So I hope we can do what I suggested above to make things go smoother.

If this sounds good to you, I'll leave you freely continue with this PR. You and ping me when you're done and need my review.

tc-mb commented:
Hi, I think there's been a bit of a misunderstanding, let me explain. I didn't expect that just over the weekend, something like this could have happened. I was busy with Omni's development and didn't see this issue and clarify it immediately. -_-

This branch is for our ongoing development; we still need to develop some features to ensure support for the high refresh rate capabilities of MiniCPM-V 4.5. My intern just wanted to see the differences from the main branch, but was a bit inexperienced, so he committed to the official llama.cpp branch.

Regarding the video interface, server support, and MiniCPM-V modifications, we added many features, and it shouldn't all be merged into one pull request. My suggestion is to wait until development is complete, confirm accuracy, and then, based on the number of added features, separately split and merge them into the llama.cpp branch.

@MrAMS As I said, we haven't finished development yet and shouldn't submit a pull request now. You can close this pull request.

@ngxson The multimodal model is starting to move towards video understanding and more features later. Although this PR was a bit of a misunderstanding, perhaps we can start considering the video interface.

We can extract the basic video capabilities and submit a simple PR. Or, as you mentioned above, you can refer to this branch and complete a PR yourself. We think both are acceptable, and we are open to hearing your opinion.

ngxson commented:
> To ensure accurate attribution, could you please add the following co-author tags to the final merge commit message?

Sure. The mentioned PR is only for demo purpose. When I upload the final PR, I will make sure to include proper Co-Author tags

ngxson commented:
Fyi @MrAMS @tc-mb , we added video support to mtmd via another implementation (see the linked issue for details)

Qwen-VL acts as an example on how input frames can be merged via a 3D convolution (temporal merge). Hope this helps start implementing the same feature on your side. Don't hesitate to open a new issue if anything is still missing

tc-mb commented:
> Fyi @MrAMS @tc-mb , we added video support to mtmd via another implementation (see the linked issue for details)
> 
> Qwen-VL acts as an example on how input frames can be merged via a 3D convolution (temporal merge). Hope this helps start implementing the same feature on your side. Don't hesitate to open a new issue if anything is still missing

ok, thanks for the reminder, I will take a look and add pr.
